src/metabase/server/middleware/session.clj

```diff
@@ -301,17 +301,33 @@
     (or (:settings (t2/select-one [User :settings] :id user-id))
         {})))
 
+(def ^:private ^:dynamic *user-local-values-user-id*
+  "User ID that we've previous bound [[*user-local-values*]] for. This exists so we can avoid rebinding it in recursive
+  calls to [[with-current-user]] if it is already bound, as this can mess things up since things
+  like [[metabase.models.setting/set-user-local-value!]] will only update the values for the top-level binding."
+  ;; placeholder value so we will end up rebinding [[*user-local-values*]] it if you call
+  ;;
+  ;;    (with-current-user nil
+  ;;      ...)
+  ;;
+  ::none)
+
 (defn do-with-current-user
-  "Impl for `with-current-user`."
+  "Impl for [[with-current-user]]."
   [{:keys [metabase-user-id is-superuser? permissions-set user-locale settings is-group-manager?]} thunk]
   (binding [*current-user-id*              metabase-user-id
             i18n/*user-locale*             user-locale
             *is-group-manager?*            (boolean is-group-manager?)
             *is-superuser?*                (boolean is-superuser?)
             *current-user*                 (delay (find-user metabase-user-id))
             *current-user-permissions-set* (delay (or permissions-set (some-> metabase-user-id user/permissions-set)))
-            *user-local-values*            (delay (atom (or settings
-                                                            (user-local-settings metabase-user-id))))]
+            ;; as mentioned above, do not rebind this to something new, because changes to its value will not be
+            ;; propagated to frames further up the stack
+            *user-local-values*            (if (= *user-local-values-user-id* metabase-user-id)
+                                             *user-local-values*
+                                             (delay (atom (or settings
+                                                              (user-local-settings metabase-user-id)))))
```

Comment:
Fixed an issue where recursive calls to `with-current-user` could break things since there would be multiple versions of the `*user-local-values*` on the stack